Tools/CISupport/ews-build/events.py

```diff
@@ -512,20 +530,26 @@ def handle_pull_request(self, payload, event):
             log.msg(f"PR #{pr_number} ({head_sha}) was updated by '{sender}', ignoring it")
             return defer.returnValue(([], 'git'))
 
-        log.msg(f'Handling PR #{pr_number} with hash: {head_sha}')
         if action == 'labeled' and GitHub.UNSAFE_MERGE_QUEUE_LABEL in labels:
             log.msg(f'PR #{pr_number} ({head_sha}) was labeled for unsafe-merge-queue')
             # 'labeled' is usually an ignored action, override it to force build
             payload['action'] = 'synchronize'
-            yield task.deferLater(reactor, self.LABEL_PROCESS_DELAY, lambda: None)
             event = 'unsafe_merge_queue'
         elif action == 'labeled' and GitHub.MERGE_QUEUE_LABEL in labels:
             log.msg(f'PR #{pr_number} ({head_sha}) was labeled for merge-queue')
             # 'labeled' is usually an ignored action, override it to force build
-            yield task.deferLater(reactor, self.LABEL_PROCESS_DELAY, lambda: None)
             payload['action'] = 'synchronize'
             event = 'merge_queue'
 
+        # We received this exact event in the last 60 seconds, ignore it
+        if cls.is_duped(f'{event}-{head_sha}'):
```

Comment:
This doesn't really seem thread-safe. If two duplicate events from GitHub come at same time, is_duped() can still return False for both of them and continue to duplicate builds.

We can either make it thread-safe. Or for urgent fix one work-around can be to move this check little later, after `yield super().handle_pull_request()`. Since that method in base class perform bunch of network requests, it would most likely add a different amount of delays for the two requests which came at same time.

Reply:
It's not thread-safe, but it doesn't need to be.

Twisted only has a single thread that's it's doing all it's work in, so as long as we only have a single web server, we can guarantee that PRs will be processed one at a time. Note that I deployed this on the staging instance and we saw exactly that happen.